[Model element] Refactor <model> element to be independent of the model process

mwyrzykowski opened the pull request:
#### 1d79eea18e89a2cd3f254414a3e8a9f040165314
<pre>
[Model element] Refactor &lt;model&gt; element to be independent of the model process
<a href="https://bugs.webkit.org/show_bug.cgi?id=299326">https://bugs.webkit.org/show_bug.cgi?id=299326</a>
<a href="https://rdar.apple.com/161127033">rdar://161127033</a>

Reviewed by Etienne Segonzac.

Add ModelPlayer for split framework, GPU process rendered model element.

Test: model-element/gpup-model-element.html

* LayoutTests/model-element/gpup-model-element-expected.png: Added.
* LayoutTests/model-element/gpup-model-element-expected.txt: Added.
* LayoutTests/model-element/gpup-model-element.html: Added.
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/Internal/WebCoreInternal.h: Added.
* Source/WebCore/Modules/Internal/module.modulemap: Added.
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.cpp:
(WebCore::DDModel::DDMeshImpl::DDMeshImpl):
(WebCore::DDModel::convertToBacking):
(WebCore::DDModel::DDMeshImpl::addMesh):
(WebCore::DDModel::DDMeshImpl::update):
(WebCore::DDModel::DDMeshImpl::render):
(WebCore::DDModel::DDMeshImpl::ioSurfaceHandles):
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.h:
* Source/WebCore/Modules/Model/InternalAPI/DDMesh.h:
(WebCore::DDModel::DDMesh::ioSurfaceHandles):
* Source/WebCore/Modules/Model/ModelDDTypes.h:
* Source/WebCore/Modules/WebGPU/GPU.cpp:
* Source/WebCore/Modules/WebGPU/GPU.h:
(WebCore::GPU::backing const):
(WebCore::GPU::backing):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUImpl.cpp:
(WebCore::WebGPU::createIOSurfaces):
(WebCore::WebGPU::GPUImpl::createModelBacking):
(WebCore::WebGPU::convertToBacking): Deleted.
(WebCore::WebGPU::GPUImpl::addMeshRequest): Deleted.
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUImpl.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPU.h:
* Source/WebCore/Modules/identity/CredentialRequestCoordinatorClient.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.h: Added.
* Source/WebCore/Modules/model-element/DDModelPlayer.mm: Added.
(makeVectorElement):
(WebCore::DDModelPlayer::create):
(WebCore::DDModelPlayer::DDModelPlayer):
(WebCore::m_page):
(WebCore::DDModelPlayer::~DDModelPlayer):
(WebCore::DDModelPlayer::load):
(WebCore::DDModelPlayer::sizeDidChange):
(WebCore::DDModelPlayer::layer):
(WebCore::DDModelPlayer::layerHostingContextIdentifier):
(WebCore::DDModelPlayer::enterFullscreen):
(WebCore::DDModelPlayer::handleMouseDown):
(WebCore::DDModelPlayer::handleMouseMove):
(WebCore::DDModelPlayer::handleMouseUp):
(WebCore::DDModelPlayer::getCamera):
(WebCore::DDModelPlayer::setCamera):
(WebCore::DDModelPlayer::isPlayingAnimation):
(WebCore::DDModelPlayer::setAnimationIsPlaying):
(WebCore::DDModelPlayer::isLoopingAnimation):
(WebCore::DDModelPlayer::setIsLoopingAnimation):
(WebCore::DDModelPlayer::animationDuration):
(WebCore::DDModelPlayer::animationCurrentTime):
(WebCore::DDModelPlayer::setAnimationCurrentTime):
(WebCore::DDModelPlayer::hasAudio):
(WebCore::DDModelPlayer::isMuted):
(WebCore::DDModelPlayer::setIsMuted):
(WebCore::DDModelPlayer::updateScene):
(WebCore::DDModelPlayer::accessibilityChildren):
(WebCore::DDModelPlayer::identifier const):
(WebCore::DDModelPlayer::displayBuffer const):
(WebCore::DDModelPlayer::contentsDisplayDelegate):
(WebCore::DDModelPlayer::currentModel):
(WebCore::DDModelPlayer::nextTexture):
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::setSourceURL):
(WebCore::HTMLModelElement::createModelPlayer):
(WebCore::HTMLModelElement::reloadModelPlayer):
(WebCore::HTMLModelElement::displayBuffer const):
(WebCore::HTMLModelElement::contentsDisplayDelegate):
(WebCore::HTMLModelElement::didFinishEnvironmentMapLoading):
(WebCore::HTMLModelElement::supportsStageModeInteraction const):
(WebCore::HTMLModelElement::attributeChanged):
(WebCore::HTMLModelElement::setEnvironmentMap):
(WebCore::HTMLModelElement::updateEnvironmentMap):
(WebCore::HTMLModelElement::environmentMapRequestResource):
(WebCore::HTMLModelElement::environmentMapResetAndReject):
(WebCore::HTMLModelElement::environmentMapResourceFinished):
(WebCore::HTMLModelElement::isURLAttribute const):
(WebCore::HTMLModelElement::memoryCost const):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.cpp:
(WebCore::PlaceholderModelPlayer::PlaceholderModelPlayer):
(WebCore::PlaceholderModelPlayer::setHasPortal):
(WebCore::PlaceholderModelPlayer::displayBuffer const):
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.h:
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.cpp:
(WebCore::DummyModelPlayer::displayBuffer const):
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/UsdModelLoader.swift:
(Converter.webUpdateMeshRequestFromUpdateMeshRequest(_:)):
* Source/WebCore/page/Navigator.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebGPU/WebGPU/DDMesh.h:
(WebGPU::DDMesh::create):
* Source/WebGPU/WebGPU/DDMesh.mm:
(WebGPU::Instance::createModelBacking):
(WebGPU::DDMesh::DDMesh):
(WebGPU::DDMesh::texture const):
(WebGPU::convertFloats):
(WebGPU::convertPart):
(WebGPU::convertUint8s):
(WebGPU::convertDescriptor):
(WebGPU::DDMesh::render const):
(WebGPU::DDMesh::update):
(WebGPU::DDMesh::addMesh):
(wgpuDDMeshAdd):
(wgpuDDMeshRender):
(WebGPU::Instance::createMesh): Deleted.
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Instance.mm:
(WebGPU::Instance::device const):
(wgpuDDMeshCreate):
(WebGPU::Instance::updateModel): Deleted.
(WebGPU::convertFloats): Deleted.
(WebGPU::convertPart): Deleted.
(): Deleted.
(WebGPU::convertUint8s): Deleted.
(WebGPU::convertDescriptor): Deleted.
(WebGPU::Instance::updateMesh): Deleted.
* Source/WebGPU/WebGPU/UsdModelRenderer.swift:
(DDBridgeReceiver.render(with:)):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
(WebKit::RemoteDDMesh::addMesh):
(WebKit::RemoteDDMesh::render):
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::RemoteGPU::createModelBacking):
(WebKit::RemoteGPU::addMeshRequest): Deleted.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
(WebKit::DDModel::RemoteDDMeshProxy::addMesh):
(WebKit::DDModel::RemoteDDMeshProxy::render):
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
(WebKit::RemoteGPUProxy::createModelBacking):
(WebKit::RemoteGPUProxy::addMeshRequest): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.h:
* Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp:
(WebKit::WebModelPlayerProvider::createModelPlayer):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::resetPreferencesToConsistentValues):

Canonical link: <a href="https://commits.webkit.org/301237@main">https://commits.webkit.org/301237@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5bdae20a6f38e38d20f188fa81e099adcdb6527a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125305 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44970 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35711 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132154 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77174 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127182 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45658 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53537 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95402 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128259 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36465 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112058 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75942 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35364 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30227 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75631 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/117395 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106233 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30453 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134840 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/123823 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52109 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39889 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103875 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52547 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108280 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103636 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26401 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48997 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27291 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49217 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52004 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57783 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/156842 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51362 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39274 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54717 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53055 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

etiennesegonzac commented on Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in:
```diff
@@ -26,12 +26,18 @@
     DispatchedFrom=WebContent,
     DispatchedTo=GPU,
-    EnabledBy=WebGPUEnabled
+#if ENABLE(GPUP_MODEL)
```
This is tripping up the parser for me (having the `#if` inside of this list).

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.h:
```diff
@@ -0,0 +1,112 @@
+#include <wtf/URL.h>
+#include <wtf/WeakPtr.h>
+
+#if ENABLE(GPUP_MODEL)
```
The whole file is gated by `ENABLE(GPUP_MODEL)` already.

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.h:
```diff
@@ -0,0 +1,112 @@
+    static Ref<DDModelPlayer> create(Page&, ModelPlayerClient&);
+    virtual ~DDModelPlayer();
+
+#if ENABLE(MODEL_PROCESS) || ENABLE(GPUP_MODEL)
```
Ditto.

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.h:
```diff
@@ -0,0 +1,112 @@
+
+    WeakPtr<ModelPlayerClient> m_client;
+
+#if ENABLE(MODEL_PROCESS) || ENABLE(GPUP_MODEL)
```
Same.

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.h:
```diff
@@ -0,0 +1,112 @@
+#if ENABLE(MODEL_PROCESS) || ENABLE(GPUP_MODEL)
+    WebCore::ModelPlayerIdentifier m_id;
+#endif
+#if ENABLE(GPUP_MODEL)
```
Same.

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,313 @@
+
+DDModelPlayer::DDModelPlayer(Page& page, ModelPlayerClient& client)
+: m_client { client }
+#if ENABLE(MODEL_PROCESS) || ENABLE(GPUP_MODEL)
```
Already in a `ENABLE(GPUP_MODEL)` block.

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,313 @@
+#if ENABLE(MODEL_PROCESS) || ENABLE(GPUP_MODEL)
+, m_id { ModelPlayerIdentifier::generate() }
+#endif
+#if ENABLE(GPUP_MODEL)
```
Same.

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,313 @@
+, m_page(page)
+#endif
+{
+    UNUSED_PARAM(page);
```
So page isn't unused :)

etiennesegonzac commented on LayoutTests/model-element/gpup-model-element.html:
```diff
@@ -0,0 +1,29 @@
+<body>
+    <h1>3D Model Viewer</h1>
+    <!-- BlendyLogo.usd ball_soccerball_realistic.usdz -->
+    <model name="model" src="file:///Volumes/Data/BlendyLogo.usd" width="600" height="400" controls="true" entityTransform="translate3d(0px, 0px, 107px)">
```
Should we commit a small uncompressed usd for this test?

mwyrzykowski replied:
Good idea, BlendyLogo is only ~3MB

mwyrzykowski replied:
Actually maybe I should just reuse cube.usdz which is already checked in for this simple test

etiennesegonzac commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -1326,8 +1326,12 @@ bool RenderLayerBacking::updateConfiguration(const RenderLayer* compositingAnces
 #endif
-        else if (auto model = element->model())
+        else if (auto model = element->model()) {
             m_graphicsLayer->setContentsToModel(WTFMove(model), element->isInteractive() ? GraphicsLayer::ModelInteraction::Enabled : GraphicsLayer::ModelInteraction::Disabled);
```
Do we want to call `setContentsToModel` ?

mwyrzykowski replied:
It doesn't seem necessary, should we conditionalize the call?

`setContentsToModel` does call `contentsLayer->setUserInteractionEnabled(interactive == ModelInteraction::Enabled);` which eventually would seem to be needed, but not right now and it could be called separately

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,313 @@
+
+        layer.setNeedsDisplay();
+
+        if (RefPtr<DDModelPlayer> player = m_modelPlayer.ptr()) {
```
Should we add a bunch of null checks and smart pointers on the stack for the player calls?

mwyrzykowski replied:
Do we need null checks for a `WeakRef`?

etiennesegonzac replied:
I meant the code inside of this block :)

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,313 @@
+    if (!m_modelLoader) {
+        auto document = corePage->localTopDocument();
+        auto window = document->window();
+        auto gpu = window->protectedNavigator()->gpu();
```
null checks?

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,313 @@
+            if (auto* machSendRight = player->nextTexture())
+                setDisplayBuffer(*machSendRight);
+
+            player->client()->element()->graphicsLayer()->setNeedsDisplay();
```
Do we need to import GraphicsLayer.h?

etiennesegonzac commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -190,6 +191,9 @@ class HTMLModelElement final : public HTMLElement, private CachedRawResourceClie
     bool isIntersectingViewport() const final { return m_isIntersectingViewport; }
 
     WEBCORE_EXPORT String modelElementStateForTesting() const;
+    WEBCORE_EXPORT const MachSendRight* displayBuffer() const;
```
nit: let's group them with the `std::optional<PlatformLayerIdentifier> layerID() const;` declaration.

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,313 @@
+            if (auto* machSendRight = player->nextTexture())
+                setDisplayBuffer(*machSendRight);
+
+            player->client()->element()->graphicsLayer()->setNeedsDisplay();
```
Actually this feels a bit cyclical...
Could we:
* have a (`render()` ?) method on the DDModelPlayer to do the DDMesh render / setDisplayBuffer part
* make sure `RenderModel::updateFromElement` / `ContentChangeType::Model` properly triggers `setNeedsDisplay` when using GPUP_MODEL
?

mwyrzykowski replied:
That sounds better - I will give that a try 👍

mwyrzykowski replied:
something like I did in the update or were you thinking of something different?

etiennesegonzac replied:
Getting closer :)

I feel pretty strongly that the `DDModelPlayer` should not reach into`HTMLModelElement` (except via the `ModelPlayerClient` interface).

And I'm curious to see `ModelDisplayBufferDisplayDelegate` could loose the `DDModelPlayer` ref if the updates were driven by the `WebUSDModelLoader` callbacks (but might be missing a detail there).

etiennesegonzac commented on Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
```diff
@@ -26,7 +26,7 @@
 #include "RemoteDDMesh.h"
 
-#if ENABLE(GPU_PROCESS)
+#if ENABLE(GPUP_MODEL)
```
nit: I think we need to update the comment after the `#endif` too

etiennesegonzac commented on Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h:
```diff
@@ -25,7 +25,7 @@
 #pragma once
 
-#if ENABLE(GPU_PROCESS)
+#if ENABLE(GPUP_MODEL)
```
Same.

etiennesegonzac commented on Source/WebCore/rendering/RenderModel.cpp:
```diff
@@ -67,6 +70,19 @@ void RenderModel::update()
         return;
     
     contentChanged(ContentChangeType::Model);
+#if ENABLE(GPUP_MODEL)
```
Oh I meant adding an ENABLE(GPUP_MODEL) block here:
https://github.com/WebKit/WebKit/blob/e6309789e815c7d4da7c3b59395bcbd205a37a1b/Source/WebCore/rendering/RenderLayerBacking.cpp#L3468-L3471
Presumably something like:
```cpp
#if ENABLE(GPUP_MODEL)
    if (m_graphicsLayer && m_graphicsLayer->drawsContent())
        m_graphicsLayer->setNeedsDisplay();
#else
   compositor().scheduleCompositingLayerUpdate();
#endif
    return;
```

(not sure about the `drawsContent` bit)

mwyrzykowski replied:
With this change:
```
RenderLayerBacking::updateConfiguration
```
never seems to get called for the model layer and then the `ModelDisplayBufferDisplayDelegate` never gets created. But:
```
#if ENABLE(GPUP_MODEL)
    if (m_graphicsLayer && m_graphicsLayer->drawsContent())
        m_graphicsLayer->setNeedsDisplay();
#endif
    compositor().scheduleCompositingLayerUpdate();
    return;
```
seems fine

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,319 @@
+
+void DDModelPlayer::load(Model& modelSource, LayoutSize size)
+{
+    UNUSED_PARAM(modelSource);
```
?

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,319 @@
+    RefPtr corePage = m_page.get();
+    m_modelLoader = nil;
+    if (!m_modelLoader) {
+        auto document = corePage->localTopDocument();
```
`RefPtr document =`

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,319 @@
+        if (!document)
+            return;
+
+        auto window = document->window();
```
`RefPtr window =`

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,319 @@
+        if (!window)
+            return;
+
+        auto gpu = window->protectedNavigator()->gpu();
```
`RefPtr gpu =`

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,319 @@
+        });
+
+        m_modelLoader = adoptNS([[WebUSDModelLoader alloc] init]);
+        auto nsURL = modelSource.url().createNSURL();
```
`RetainPtr nsURL =`

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,319 @@
+
+        layer.setNeedsDisplay();
+
+        if (RefPtr player = m_modelPlayer.get(); player->currentModel())
```
Oh maybe I'm missing something, but I thought `DDModelPlayer::update()` could be triggered from the modelLoader callbacks...

mwyrzykowski replied:
We need to emulate, somehow, how Canvas works so the IOSurface gets re-painted every time the page is updated.

For Canvas, `Page::doAfterUpdateRendering` does:
```
    forEachRenderableDocument([] (Document& document) {
        document.prepareCanvasesForDisplayOrFlushIfNeeded();
    });
```
Inside `Document::prepareCanvasesForDisplayOrFlushIfNeeded` each Canvas in the document gets `HTMLCanvasElement::prepareForDisplay` called which is similar to `HTMLModelElement::update` in this case.

Instead of maintaining a list of `HTMLModelElement` in the Page which seems unnecessary, I just mark the element needing display after it is displayed because <model> needs to be repainted during animation.

Perhaps eventually this should only occur if 'autoplay=true' is set or if the user is interacting with the <model> on the page. There could also be other reasons to redraw the model, such as responding to device orientation

etiennesegonzac replied:
I see...
Let's revisit once we have a clearer view on animation / interaction :)
(wonder if we'll need extra callbacks to maintain the `currentTime`/animation state for example)

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,319 @@
+        , m_isOpaque(isOpaque)
+    {
+    }
+    WeakRef<DDModelPlayer> m_modelPlayer;
```
...removing the need for this WeakRef.

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,319 @@
+                m_currentModel->addMesh(toCpp(addRequest));
+        } modelUpdatedCallback:^(WebUpdateMeshRequest *updateRequest) {
+            if (m_currentModel)
+                m_currentModel->update(toCpp(updateRequest));
```
(Maybe we need an extra `ModelPlayerClient::didUpdateRendering()` method for this.)

mwyrzykowski replied:
We can add one but it would only result in a single call

When the model is animating, none of these callbacks would be regularly called into.

etiennesegonzac replied:
Oh ok, I thought it might.

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,319 @@
+    if (auto* machSendRight = nextTexture(); machSendRight && contentsDisplayDelegate())
+        RefPtr { m_contentsDisplayDelegate }->setDisplayBuffer(*machSendRight);
+
+    if (RefPtr client = m_client.get()) {
```
For the didFinishLoading case at least this should not be needed (the client already triggers `RenderModel::updateFromElement`)...

mwyrzykowski replied:
That's correct but I think we need it when driving the display link for the IOSurface

etiennesegonzac commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1517,6 +1560,14 @@ String HTMLModelElement::modelElementStateForTesting() const
     return "Unknown"_s;
 }
 
+#if ENABLE(GPUP_MODEL)
```
...potentially letting us remove this....

etiennesegonzac commented on Source/WebCore/Modules/model-element/ModelPlayerClient.h:
```diff
@@ -63,6 +64,7 @@ class WEBCORE_EXPORT ModelPlayerClient : public CanMakeWeakPtr<ModelPlayerClient
     virtual bool isVisible() const = 0;
     virtual bool isIntersectingViewport() const = 0;
     virtual void logWarning(ModelPlayer&, const String& warningMessage) = 0;
+    virtual HTMLModelElement* element() = 0;
```
...and this...

etiennesegonzac commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -247,6 +250,7 @@ class HTMLModelElement final : public HTMLElement, private CachedRawResourceClie
     std::optional<PlatformLayerIdentifier> modelContentsLayerID() const final;
     bool isVisible() const final;
     void logWarning(ModelPlayer&, const String&) final;
+    HTMLModelElement* element() final { return this; }
```
...and this.

mwyrzykowski replied:
I wasn't sure how to drive the display link without the WeakRef so I left this line and the above two

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -0,0 +1,316 @@
+
+        layer.setNeedsDisplay();
+
+        if (RefPtr player = m_modelPlayer.get(); player->currentModel())
```
nit: we could move the `player->currentModel()` check to `DDModelPlayer::update()`.

etiennesegonzac reviewed (approved):
We can introduce a (GPUP_MODEL only) `didUpdateDisplayDelegate` on ModelClient in this patch (instead of exposing the whole HTMLModelElement). And revisit _when_ we call once we have more animation/interaction code!

mwyrzykowski committed:
[Model element] Refactor <model> element to be independent of the model process
https://bugs.webkit.org/show_bug.cgi?id=299326
rdar://161127033

Reviewed by Etienne Segonzac.

Add ModelPlayer for split framework, GPU process rendered model element.

Test: model-element/gpup-model-element.html

* LayoutTests/model-element/gpup-model-element-expected.png: Added.
* LayoutTests/model-element/gpup-model-element-expected.txt: Added.
* LayoutTests/model-element/gpup-model-element.html: Added.
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/Internal/WebCoreInternal.h: Added.
* Source/WebCore/Modules/Internal/module.modulemap: Added.
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.cpp:
(WebCore::DDModel::DDMeshImpl::DDMeshImpl):
(WebCore::DDModel::convertToBacking):
(WebCore::DDModel::DDMeshImpl::addMesh):
(WebCore::DDModel::DDMeshImpl::update):
(WebCore::DDModel::DDMeshImpl::render):
(WebCore::DDModel::DDMeshImpl::ioSurfaceHandles):
* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.h:
* Source/WebCore/Modules/Model/InternalAPI/DDMesh.h:
(WebCore::DDModel::DDMesh::ioSurfaceHandles):
* Source/WebCore/Modules/Model/ModelDDTypes.h:
* Source/WebCore/Modules/WebGPU/GPU.cpp:
* Source/WebCore/Modules/WebGPU/GPU.h:
(WebCore::GPU::backing const):
(WebCore::GPU::backing):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUImpl.cpp:
(WebCore::WebGPU::createIOSurfaces):
(WebCore::WebGPU::GPUImpl::createModelBacking):
(WebCore::WebGPU::convertToBacking): Deleted.
(WebCore::WebGPU::GPUImpl::addMeshRequest): Deleted.
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUImpl.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPU.h:
* Source/WebCore/Modules/identity/CredentialRequestCoordinatorClient.h:
* Source/WebCore/Modules/model-element/DDModelPlayer.h: Added.
* Source/WebCore/Modules/model-element/DDModelPlayer.mm: Added.
(makeVectorElement):
(WebCore::DDModelPlayer::create):
(WebCore::DDModelPlayer::DDModelPlayer):
(WebCore::m_page):
(WebCore::DDModelPlayer::~DDModelPlayer):
(WebCore::DDModelPlayer::load):
(WebCore::DDModelPlayer::sizeDidChange):
(WebCore::DDModelPlayer::layer):
(WebCore::DDModelPlayer::layerHostingContextIdentifier):
(WebCore::DDModelPlayer::enterFullscreen):
(WebCore::DDModelPlayer::handleMouseDown):
(WebCore::DDModelPlayer::handleMouseMove):
(WebCore::DDModelPlayer::handleMouseUp):
(WebCore::DDModelPlayer::getCamera):
(WebCore::DDModelPlayer::setCamera):
(WebCore::DDModelPlayer::isPlayingAnimation):
(WebCore::DDModelPlayer::setAnimationIsPlaying):
(WebCore::DDModelPlayer::isLoopingAnimation):
(WebCore::DDModelPlayer::setIsLoopingAnimation):
(WebCore::DDModelPlayer::animationDuration):
(WebCore::DDModelPlayer::animationCurrentTime):
(WebCore::DDModelPlayer::setAnimationCurrentTime):
(WebCore::DDModelPlayer::hasAudio):
(WebCore::DDModelPlayer::isMuted):
(WebCore::DDModelPlayer::setIsMuted):
(WebCore::DDModelPlayer::updateScene):
(WebCore::DDModelPlayer::accessibilityChildren):
(WebCore::DDModelPlayer::identifier const):
(WebCore::DDModelPlayer::displayBuffer const):
(WebCore::DDModelPlayer::contentsDisplayDelegate):
(WebCore::DDModelPlayer::currentModel):
(WebCore::DDModelPlayer::nextTexture):
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::setSourceURL):
(WebCore::HTMLModelElement::createModelPlayer):
(WebCore::HTMLModelElement::reloadModelPlayer):
(WebCore::HTMLModelElement::displayBuffer const):
(WebCore::HTMLModelElement::contentsDisplayDelegate):
(WebCore::HTMLModelElement::didFinishEnvironmentMapLoading):
(WebCore::HTMLModelElement::supportsStageModeInteraction const):
(WebCore::HTMLModelElement::attributeChanged):
(WebCore::HTMLModelElement::setEnvironmentMap):
(WebCore::HTMLModelElement::updateEnvironmentMap):
(WebCore::HTMLModelElement::environmentMapRequestResource):
(WebCore::HTMLModelElement::environmentMapResetAndReject):
(WebCore::HTMLModelElement::environmentMapResourceFinished):
(WebCore::HTMLModelElement::isURLAttribute const):
(WebCore::HTMLModelElement::memoryCost const):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.cpp:
(WebCore::PlaceholderModelPlayer::PlaceholderModelPlayer):
(WebCore::PlaceholderModelPlayer::setHasPortal):
(WebCore::PlaceholderModelPlayer::displayBuffer const):
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.h:
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.cpp:
(WebCore::DummyModelPlayer::displayBuffer const):
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayer.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/UsdModelLoader.swift:
(Converter.webUpdateMeshRequestFromUpdateMeshRequest(_:)):
* Source/WebCore/page/Navigator.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebGPU/WebGPU/DDMesh.h:
(WebGPU::DDMesh::create):
* Source/WebGPU/WebGPU/DDMesh.mm:
(WebGPU::Instance::createModelBacking):
(WebGPU::DDMesh::DDMesh):
(WebGPU::DDMesh::texture const):
(WebGPU::convertFloats):
(WebGPU::convertPart):
(WebGPU::convertUint8s):
(WebGPU::convertDescriptor):
(WebGPU::DDMesh::render const):
(WebGPU::DDMesh::update):
(WebGPU::DDMesh::addMesh):
(wgpuDDMeshAdd):
(wgpuDDMeshRender):
(WebGPU::Instance::createMesh): Deleted.
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Instance.mm:
(WebGPU::Instance::device const):
(wgpuDDMeshCreate):
(WebGPU::Instance::updateModel): Deleted.
(WebGPU::convertFloats): Deleted.
(WebGPU::convertPart): Deleted.
(): Deleted.
(WebGPU::convertUint8s): Deleted.
(WebGPU::convertDescriptor): Deleted.
(WebGPU::Instance::updateMesh): Deleted.
* Source/WebGPU/WebGPU/UsdModelRenderer.swift:
(DDBridgeReceiver.render(with:)):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.cpp:
(WebKit::RemoteDDMesh::addMesh):
(WebKit::RemoteDDMesh::render):
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.h:
* Source/WebKit/GPUProcess/graphics/Model/RemoteDDMesh.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::RemoteGPU::createModelBacking):
(WebKit::RemoteGPU::addMeshRequest): Deleted.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
(WebKit::DDModel::RemoteDDMeshProxy::addMesh):
(WebKit::DDModel::RemoteDDMeshProxy::render):
* Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
(WebKit::RemoteGPUProxy::createModelBacking):
(WebKit::RemoteGPUProxy::addMeshRequest): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.h:
* Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp:
(WebKit::WebModelPlayerProvider::createModelPlayer):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::resetPreferencesToConsistentValues):

Canonical link: https://commits.webkit.org/301237@main
